### PR TITLE
LPS-152906 Add Sharing Button

### PR DESCRIPTION
**Pull request by cgoncas**
### Motivation

As a content marketer, I want to view a preview of a blog image and web content in the info panel, so that I can more thoroughly understand the context of the asset at a glance.

### Proposed Solution

We have evaluated several options, all the discussion can be found in the T[echnical Refinement Document](https://liferay.atlassian.net/wiki/spaces/ENGTANGO/pages/2065990182/LPS-152906+Add+a+share+button+to+the+info+panel+of+the+content+dashboard) and https://github.com/liferay-tango/liferay-portal/pull/2602.

Finally we decided that the safest approach was to create a sharing jsp that includes the sharing:button taglib, so the sharing button is rendered using a fetch, as we do with the content performance panel.

To do so, we also needed to set the _sharingJavaScriptFactory.requestSharingJavascript(); in the ContentDashboard Portlet, because the Info Panel is a Javascript component that is rendered after the boton.jsp is loaded, so we can't set the ThreadLocal after the view.jsp is loaded.

_This solution include the PRs:_

First PR:  [Include the JS code needed so the button taglib can work](https://github.com/liferay-lima/liferay-portal/pull/3277)
Second PR: [new sharing_button JSP, that includes the sharing button taglib](https://github.com/liferay-tango/liferay-portal/pull/2622)
Third PR: [Extract download button from Preview component](https://github.com/liferay-tango/liferay-portal/pull/2612)
Fourth OR: [Implement Share Button Action in Blogs and File Entries](https://github.com/liferay-tango/liferay-portal/pull/2632)
Fifth PR: [Implement Frontend](https://github.com/liferay-tango/liferay-portal/pull/2634)
_Screencaps_

Captura de pantalla 2022-07-22 a las 16 06 47
Captura de pantalla 2022-07-22 a las 16 06 40

### How to test it

1. Enable FF in portal-ext.properties: feature.flag.LPS-152905=true

### CODE REVIEW

**Comment by cgoncas** [comment]
@claraizquierdo , could you add the code to render the Share button in the info panel with the given URL?

Let me know what do you think.

See you!!

**Comment by beltranrengifo** [comment]
hi @cgoncas !
quick question: is this implemented for documents & blogs or just for blogs?

**Comment by claraizquierdo** [comment]
> @claraizquierdo , could you add the code to render the Share button in the info panel with the given URL?
> 
> Let me know what do you think.
> 
> See you!!

Ei @cgoncas , yes, I going to start with it now.

**Comment by claraizquierdo** [comment]
Ei @cgoncas , the server is returning an error when I try to show the info bar:
Captura de pantalla 2022-07-15 a las 9 32 41

**Comment by claraizquierdo** [comment]
@cgoncas sorry, is working for blogs. I thought that is was working for documents too.

**Comment by claraizquierdo** [comment]
Hello again @cgoncas ,

I just pushed my code fetching the button from the url. As you can see, the url is returning also the top menu bar and I think it shouldn't. Also, the button doesn't do anything if I click on it.

**Comment by cgoncas** [comment]
Hey @claraizquierdo !!

What do you think is missing in the button, so it doesn't execute any action?

Thanks!!

File: modules/apps/content-dashboard/content-dashboard-blogs-impl/src/main/java/com/liferay/content/dashboard/blogs/internal/item/action/SharingBlogsEntryContentDashboardItemAction.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	public String getURL() {
+		try {
+			PortletResponse portletResponse =
+				(PortletResponse)_httpServletRequest.getAttribute(
+					JavaConstants.JAVAX_PORTLET_RESPONSE);
+
+			LiferayPortletResponse liferayPortletResponse =
+				_portal.getLiferayPortletResponse(portletResponse);
+
+			PortletURL portletURL = PortletURLBuilder.createRenderURL(
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-blogs-impl/src/main/java/com/liferay/content/dashboard/blogs/internal/item/action/SharingBlogsEntryContentDashboardItemAction.java:81]: Hey @claraizquierdo !!

The URL is being generated here, could you add the pop_up parameter so we don't have the top menu?

Thanks!!

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-blogs-impl/src/main/java/com/liferay/content/dashboard/blogs/internal/item/action/SharingBlogsEntryContentDashboardItemAction.java:82]: Awesome!!!, thank you :)

**Comment by cgoncas** [comment]
> The dropdown menu will not work with this solution as it is, now we are redirecting the user to a new page with a button. I think we should use this [`createShareDropdownItem`](https://github.com/liferay/liferay-portal/blob/f30469833a60c05cf683b3422632f1bdfe43c472/modules/apps/sharing/sharing-web/src/main/java/com/liferay/sharing/web/internal/display/context/util/SharingMenuItemFactoryImpl.java#L108) in this case
> 
> Does it make sense to have two types in `ContentDashboardItemAction`, like SHARING_BUTTON and SHARING_MENU_ITEM?
> 
> image
> 
> And we should add new checks, as I see [here](https://github.com/liferay/liferay-portal/blob/4af8937052647e8f1913d3e8c8702f994252f4cb/modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/servlet/taglib/util/BlogsEntryActionDropdownItemsProvider.java#L91) 😢

Yeahh Beltrán, as we do with the view-in-panel, if we want to add a dropdown button we should use a different action, but if we do that, it will be in another issue.

**Comment by beltranrengifo** [comment]
> > The dropdown menu will not work with this solution as it is, now we are redirecting the user to a new page with a button. I think we should use this [`createShareDropdownItem`](https://github.com/liferay/liferay-portal/blob/f30469833a60c05cf683b3422632f1bdfe43c472/modules/apps/sharing/sharing-web/src/main/java/com/liferay/sharing/web/internal/display/context/util/SharingMenuItemFactoryImpl.java#L108) in this case
> > Does it make sense to have two types in `ContentDashboardItemAction`, like SHARING_BUTTON and SHARING_MENU_ITEM?
> > image
> > And we should add new checks, as I see [here](https://github.com/liferay/liferay-portal/blob/4af8937052647e8f1913d3e8c8702f994252f4cb/modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/servlet/taglib/util/BlogsEntryActionDropdownItemsProvider.java#L91) 😢
> 
> Yeahh Beltrán, as we do with the view-in-panel, if we want to add a dropdown button we should use a different action, but if we do that, it will be in another issue.

Good @cgoncas ! Then I stop my dropdown item issue exploration here and will discuss later how to solve it. Thxs!

**Review by beltranrengifo — COMMENTED** [review]
Some comments to the JS part 😺

**Comment by Stephy6** [comment]
Hi @cgoncas! 
I added the files I think we need to add the Share button to the Info Panel for Documents, however, the Info Panel shows this error when I click over the Info Button for Documents (the request for the button isn't made). Could you help me? 

cc: @beltranrengifo 

Screenshot 2022-07-19 at 10 52 15

File: modules/apps/content-dashboard/content-dashboard-document-library-impl/src/main/java/com/liferay/content/dashboard/document/library/internal/item/action/SharingFileEntryContentDashboardItemAction.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+				_portal.getLiferayPortletResponse(portletResponse);
+
+			PortletURL portletURL = PortletURLBuilder.createRenderURL(
+				liferayPortletResponse,
+				"com_liferay_content_dashboard_web_portlet_" +
+				"ContentDashboardAdminPortlet"
+			).setMVCPath(
+				"/sharing_button.jsp"
+			).setParameter(
+				"className", FileEntry.class.getName()
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-document-library-impl/src/main/java/com/liferay/content/dashboard/document/library/internal/item/action/SharingFileEntryContentDashboardItemAction.java:89]: Hey @Stephy6 !!!

Have you tried to use DLFileEntry.class.getName() here?

Thanks!!

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-document-library-impl/src/main/java/com/liferay/content/dashboard/document/library/internal/item/action/SharingFileEntryContentDashboardItemAction.java:89]: Awesome!!, do you have any problem in the Logs?

Thanks :)

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 3]
```diff
@@ -151,6 +151,9 @@ protected void doServeResource(
 					).orElse(
 						null
 					)
+				).put(
+					"sharingURL",
```
[the remark this answers is not in the document]
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:155]: Great!!

With the content performance panel we use fetchURL, so I think fetchSharingButtonURL is a very good name, should I rename it right now @claraizquierdo and @beltranrengifo !!

**Comment by cgoncas** [comment]
> Ei @cgoncas , the server is returning an error when I try to show the info bar:

So, no erros in the server, right?, in the server console logs, you don't see any errors, right?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:155]: Yeah please!

**Comment by beltranrengifo** [comment]
> Hey @claraizquierdo and @beltranrengifo !!!
> 
> Could you test if this PR is loading the JS code when it is needed?
> 
> Thanks!!

On my way

**Comment by cgoncas** [comment]
> > > Ei @cgoncas , the server is returning an error when I try to show the info bar:
> > 
> > 
> > So, no erros in the server, right?, in the server console logs, you don't see any errors, right?
> 
> No errors in server console, nor in browser's one, just a normal 200 response with `{"error":"An unexpected error occurred."}` payload

Awesome!!!

On my way to debug it :)

**Comment by claraizquierdo** [comment]
> Hey @claraizquierdo !!
> 
> maybe we can send to master the code that extract the download action from the Preview componente, right?
> 
> Thanks!!

Yes @cgoncas , you are right. Let me open a PR with just that

**Comment by cgoncas** [comment]
Hey @Stephy6,  @beltranrengifo  and @claraizquierdo !!!

I am going to force push this branch, so I can separate the commits of the JSP, and send it to master

Thanks!!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java]: Done, thank you :(

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+						elRef.current.innerHTML = html;
+					})
+				)
+				.catch((error) => {
+					console.log('Failed to fetch share button: ', error);
+				});
+		}
+	}, [sharingURL]);
+
+	return <div ref={elRef} />;
```
[the remark this answers is not in the document]
Comment (claraizquierdo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js]: Fixed

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		if (elRef.current) {
+			fetch(sharingURL, {
+				method: 'GET',
+			})
+				.then((response) =>
+					response.text().then((html) => {
+						elRef.current.innerHTML = html;
+					})
+				)
+				.catch((error) => {
```
[the remark this answers is not in the document]
Comment (claraizquierdo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js:33]: Let me check with the design team

[the remark this answers is not in the document]
Comment (claraizquierdo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js:33]: Asked here: https://issues.liferay.com/browse/LPS-152906?focusedCommentId=2724540&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2724540

**Comment by Stephy6** [comment]
Hi @cgoncas and @beltranrengifo, I've just added the changes to display the Share button but its displayed like this: 
Screenshot 2022-07-19 at 19 27 27

**Comment by beltranrengifo** [comment]
> Hey @claraizquierdo and @beltranrengifo !!!
> 
> Could you test if this PR is loading the JS code when it is needed?
> 
> Thanks!!

@cgoncas @claraizquierdo I cannot reproduce the FE JS dependency error
I'm not 100% sure about this as I had not a consistent way to make it fail, but seems better anyway. I'll let you know 🙌

**Comment by beltranrengifo** [comment]
> Hi @cgoncas and @beltranrengifo, I've just added the changes to display the Share button but its displayed like this: Screenshot 2022-07-19 at 19 27 27

That looks good @Stephy6 !
Are this changes in the pr already? I can take a look to the styling

**Comment by claraizquierdo** [comment]
> > Hi @cgoncas and @beltranrengifo, I've just added the changes to display the Share button but its displayed like this: Screenshot 2022-07-19 at 19 27 27
> 
> That looks good @Stephy6 ! Are this changes in the pr already? I can take a look to the styling

@Stephy6 I fixed that today

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+const Share = ({sharingURL}) => {
+	const elRef = useRef(null);
+
+	useEffect(() => {
+		if (elRef.current) {
+			fetch(sharingURL, {
+				method: 'GET',
+			})
+				.then((response) =>
+					response.text().then((html) => {
```
[the remark this answers is not in the document]
Comment (claraizquierdo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js]: Thank you for pointing that out, I like being cautious. I just pushed these changes.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+const Share = ({sharingURL}) => {
+	const elRef = useRef(null);
+
+	useEffect(() => {
+		if (elRef.current) {
+			fetch(sharingURL, {
+				method: 'GET',
```
[the remark this answers is not in the document]
Comment (claraizquierdo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/SidebarPanelInfoView/Share.js]: Done

**Comment by Stephy6** [comment]
Hi @cgoncas and @beltranrengifo  ! I think is working fine now, what do you think?
Thank u!

**Comment by claraizquierdo** [comment]
The frontend commits are now organized @beltranrengifo, @Stephy6 and @cgoncas

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by Stephy6** [comment]
Hi @cgoncas! I've added the tests related to the Share button, let me know what you think!

**Comment by cgoncas** [comment]
Closing this PR, because the only missing code are the tests, and will be send in https://github.com/liferay-tango/liferay-portal/pull/2648

**Pull request closed without merging**